File: problem3.java
//Time Complexity: O(n log k), where n is the size of the input array nums.
//Space Complexity: O(n), where n is the size of the input array nums.
//ran on leetcode: Yes
// use a priority queue and a hash map to find the k most frequent elements in an array
class Element{
    int val, freq;
    Element(int val, int freq)
    {
        this.val = val;
        this.freq = freq;
    }
}
class Solution {
    public int[] topKFrequent(int[] nums, int k) {
        PriorityQueue<Element> pq = new PriorityQueue<>((a,b)->a.freq-b.freq);
        HashMap<Integer,Integer> hm = new HashMap<>();
        
        for(int i=0;i<nums.length;i++)
        {
            hm.put(nums[i], hm.getOrDefault(nums[i],0)+1);
        }
        
        for(Integer key : hm.keySet())
        {
            pq.add(new Element((int)key, hm.get(key)));
            if(pq.size()>k)
                pq.remove();
        }
        
        int[] result = new int[k];
        for(int i=0;i<k;i++)
        {
            result[i] = pq.remove().val;
        }
        return result;
    }
}

File: Problem3.py
class Solution:
    def topKFrequent(self, nums, k):
        # Time complexity -O(n log k), min heap and extracting the k most frequent elements each take O(n log k), where k is the number of elements we want to find.
        # Space complexity - O(n + k), which is dominated by the O(n) space used for the count dictionary.
        # Step 1: Count the frequency of each element using a Counter
        count = Counter(nums)

        # Step 2: Create a min heap to store the k most frequent elements
        min_heap = []

        for num, freq in count.items():
            # Push elements into the min heap
            heapq.heappush(min_heap, (freq, num))

            # If the size of the min heap exceeds k, remove the smallest element
            if len(min_heap) > k:
                heapq.heappop(min_heap)

        # Step 3: Extract the k most frequent elements from the min heap
        result = []
        while min_heap:
            result.append(heapq.heappop(min_heap)[1])

        # Reverse the result to get elements in descending order of frequency
        result.reverse()

        return result
